pkg/agent/pipeline.go

```diff
@@ -148,21 +148,25 @@ func (s *NimbessPipeline) createPipeline(module network.PipelineModule, port str
 		// We can assert Module safely here because every complex module should always inherit Module
 		newMod.(network.PipelineModule).SetName(newName)
 		// Create new maps for gates, as new mappings will be built recursively
-		newMod.(network.PipelineModule).SetEGateMap(make(map[network.Gate]network.PipelineModule))
-		newMod.(network.PipelineModule).SetIGateMap(make(map[network.Gate]network.PipelineModule))
+		newMod.(network.PipelineModule).SetEGateMap(network.MakeGateMap())
+		newMod.(network.PipelineModule).SetIGateMap(network.MakeGateMap())
 		// Append new module to existing module list
 		s.Modules = append(s.Modules, newMod.(network.PipelineModule))
 		log.Debugf("Module created: %+v", newMod)
 	}
 
 	// Walk meta module Ingress gates and connect new port pipeline modules
 	for gate, mod := range module.GetIGateMap() {
-		newMod.Connect(s.createPipeline(mod, port), false, &gate)
+		if mod != nil {
```

Comment:
I'm a little wary of this behavior where it is ok that a module is nil. The reason why is that it will be hard to differentiate if we are intentionally having a mod as nil (reserved gate) or if there is a real bug in the code. The previous way the code worked is that there shouldn't be any nil pointer. I'm thinking it might be better to make a new Nimbess Module type called Reserved and use that to represent an intentional act to reserve the gate. Then in all these places we check to see if mod == reflect.TypeOf(&modules.Reserved{}). I think it is still also good to keep the checks for nil, and raise an error if a mod is nil. Will help debug issues in the future.

Reply:
I am reaching the same conclusion.

However instead of Reserved, I am thinking to use here a sink instance to implement a default "drop if you do not know what you are doing"